src/chain/auth.sw

```diff
@@ -1,41 +1,160 @@
 library auth;
-//! Functionality for determining who is calling an ABI method
+//! Functionality for determining who is calling a contract.
 
-use ::result::Result;
 use ::address::Address;
+use ::b512::B512;
+use ::chain::assert;
 use ::contract_id::ContractId;
+use ::result::Result;
+
 
 pub enum AuthError {
     ContextError: (),
 }
 
 pub enum Sender {
     Address: Address,
-    Id: ContractId,
+    ContractId: ContractId,
 }
 
-/// Returns `true` if the caller is external (ie: a script or predicate).
-// ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#gm-get-metadata
+/// Returns `true` if the caller is external (i.e. a script).
+/// ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#gm-get-metadata
 pub fn caller_is_external() -> bool {
     asm(r1) {
         gm r1 i1;
         r1: bool
     }
 }
 
-/// Get the `Sender` (ie: `Address`| ContractId) from which a call was made.
-/// Returns a Result::Ok(Sender) or Result::Error.
-// NOTE: Currently only returns Result::Ok variant if the parent context is Internal.
+fn caller_contract_id() -> ContractId {
+    ~ContractId::from(asm(r1) {
+        gm r1 i2;
```

Comment:
I'm not 100% here, but in the case of calling a contract from a script, would this (`GM::get_caller`) be the transaction ID rather than a `ContractId`?
Ref: https://fuellabs.slack.com/archives/C01L1HUJSL8/p1645190705243929?thread_ts=1645133890.080069&cid=C01L1HUJSL8

Reply:
If so, then the `Sender` struct should perhaps keep the second variant name more generic, ie: `Id` ?

Reply:
cc @adlerjohn This is something I'm trying to debug currently but is also relevant to this thread.
So, as a sanity check:

Let's say  I call a contract from the SDK, and the contract uses the `gm` opcode (ie: the `msg_sender()` function).
In the case where `imm == GM_GET_CALLER` , I would expect 2 things:
1. calling `gm` here should **not** panic, and
2. $rA will point to the previous call frame's contract ID (which here is actually a `TransactionId` I think.

ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#gm-get-metadata

Reply:
If you call a contract from the SDK, then the caller of that contract is a script. So `GM` with `GM_GET_CALLER` should panic, because there's no contract calling.